typing/types.ml

```diff
@@ -135,36 +135,70 @@ and method_privacy =
   | Mprivate of field_kind
 
 (* Variance *)
+(* Variance forms a product lattice of the following partial orders:
+     0 <= may_pos <= pos
+     0 <= may_weak <= may_neg <= neg
+     0 <= inj
+   Additionally, the following implications are valid
+     pos => inj
+     neg => inj
+   Examples:
+     type 'a t        : may_pos + may_neg + may_weak
+     type 'a t = 'a   : pos
+     type 'a t = 'a -> unit : neg
+     type 'a t = ('a -> unit) -> unit : pos + may_weak
+     type 'a t = A of (('a -> unit) -> unit) : pos
+     type +'a p = ..  : may_pos + inj
+     type +!'a t      : may_pos + inj
+     type -!'a t      : may_neg + inj
+     type 'a t = A    : inj
+ *)
 
 module Variance = struct
   type t = int
   type f = May_pos | May_neg | May_weak | Inj | Pos | Neg | Inv
```

Comment:
What is `Inv`, again?

Reply:
Just the combination of `Pos` and `Neg`.
Previously there was an extra bit, but a deeper analysis made clear that this was not necessary. I just keep it as it simplifies the code.
Actually one could argue that `Pos` and `Neg` are not really necessary either, because the only information that really matters is `Inj`, and remembering that indices of GADTs are invariant (using `Inv o Inj = Inv`). Yet this externalization of `Inv` is a bit confusing.

Reply:
Ah yes of course, "inv" for Invariant. Sorry for the silly questions.) If you removed `Inv`, then the type `f` would be exactly the element of the orders you specified. (We would express `single Inv` with `union (single Pos) (single Neg)`.) Would this not be a simplification?

(Then you suggest to get rid of `Pos` and `Neg` and replace them by `May_pos + Inj`, `May_neg + Inj`, I'm not sure about that; they behave similarly in the lattice but they don't really mean the same thing, `Pos` is more information than `May_pos + Inj` and we might want to distinguish them again in the future.)